### PR TITLE
fix: Fix the definition of Operation

### PR DESCRIPTION
**Pull request by Jason3S**
An operation is a function that takes an Iterator and returns an Iterator.

But this was begin hidden by the `any`s in the `pipe` function.

**tldr;**
Many years ago before async iterators existed, I wrote [GenSequence: Small library to simplify working with Generators and Iterators in JavaScript / Typescript](https://github.com/Jason3S/GenSequence). It is a bit over done. Too much Iterator conversion going on to support chaining. For a while I wanted to simplify and drop the chaining by moving towards using a `pipe` approach while reducing overhead.

I recently saw your implementation and considered deprecating GenSequence. But I noticed that this repo uses a lot of `any`s, which hide possible problems.

For example, Operation's definition didn't match how it was being used in `pipe`, which is the reason for this PR.

If you would like some help removing the `any`s, I can help. I do suggest making both `pipeSync` and `pipeAsync` functions. This will simplify things.

### CODE REVIEW

**Comment by RebeccaStevens** [comment]
Hey @Jason3S 

I'm currently working on refactoring this library to remove all those `any` :)
I just discovered the same issue as this PR tries to resolve last night and I like your approach. I do have one change I'd recommend though (I'll comment it on the code snippet).

I also agree that `pipe` should be broken up into two function `pipeSync` (or maybe this just stays as `pipe`) and `pipeAsync`. This not only will make tying the function a lot easier but will also allow uses to explicitly specify what type of pipeline they want.
Currently I believe it is not possible to create an async pipeline starting with a sync iterable.

File: src/types.ts [hunk 1]
```diff
@@ -109,7 +109,8 @@ export type AnyIterableIterator<T> = AnyIterable<T> | AnyIterator<T>;
  * Pipe-through type (return type for all operators)
  */
 export interface Operation<T, R> {
-    (i: AnyIterable<T>): Operation<T, R>;
+    (i: Iterable<T>): Iterable<R>;
+    (i: AsyncIterable<T>): AsyncIterable<R>;
```
Comment (RebeccaStevens) [inline src/types.ts:113]: This is specifying that an operation must support both sync and async behaviors.
Instead it should state that it must support at least one.

This could possible be the definition:

```ts
export type Operation<T, R> = {
    (i: Iterable<T>): Iterable<R>;
} | {
    (i: AsyncIterable<T>): AsyncIterable<R>;
};
```

Though it would probably be better to break that down further into `OperationSync`, `OperationAsync` and `AnyOperation`.

Comment (Jason3S) [reply src/types.ts:113]: @RebeccaStevens,

> Though it would probably be better to break that down further into OperationSync, OperationAsync and AnyOperation.

Along the same lines, I had made the following change, but rejected it to keep the diff smaller:
```ts
export type Operation<T, R>  = OperationSync<T, R> | OperationAsync<T, R>;

export interface OperationSync<T, R> {
    (i: Iterable<T>): Iterable<R>;
}

export interface OperationAsync<T, R> {
    (i: AsyncIterable<T>): AsyncIterable<R>;
}
```

**Comment by Jason3S** [comment]
> I'm currently working on refactoring this library to remove all those `any` :) 

Wonderful! Please feel free to close this PR.

> I also agree that `pipe` should be broken up into two function `pipeSync` (or maybe this just stays as `pipe`) and `pipeAsync`. This not only will make tying the function a lot easier but will also allow uses to explicitly specify what type of pipeline they want. 

I agree, I think it is better to keep *sync* and *async* separate until there is a real need to combine them.

> Currently I believe it is not possible to create an async pipeline starting with a sync iterable.
The current `pipe` method does not support it.

**Comment by vitaly-t** [comment]
@RebeccaStevens Do you think this PR is still relevant, or should it be closed?

CC: @Jason3S

**Comment by RebeccaStevens** [comment]
These changes have already been made so it's safe to close this PR.

**Comment by vitaly-t** [comment]
Closing, as the changes have already been made.
